File: pyround1/34.在排序数组中查找元素的第一个和最后一个位置.py
#
# @lc app=leetcode.cn id=34 lang=python
#
# [34] 在排序数组中查找元素的第一个和最后一个位置
#

# @lc code=start
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        left, right = 0, len(nums) -1 
        if not nums or target > nums[right] or target < nums[left]:
            return [-1, -1]
        while left < right:
            while nums[left] < target:
                left += 1
            while nums[right] > target:
                right -= 1
            if nums[left] == nums[right]:
                break
        return [left, right] if left <= right else [-1, -1]
        
        
# @lc code=end

File: cpprounttmp/34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */

// @lc code=start
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        if (nums.empty())
            return {-1, -1};
        int firstPosition = findFirstPosition(nums, target);
        if (firstPosition == -1)
            return {-1, -1};
        int lastPosition = FindLastPosition(nums, target);
        return vector<int>{firstPosition, lastPosition};
    }
private:
    int findFirstPosition(vector<int> &nums, int target){
        int left = 0;
        int right = nums.size() - 1;
        while(left <= right){
            int mid = (left + right) / 2;
            if(nums[mid] == target)
                right = mid - 1;
            else if (nums[mid] < target)
                left = mid + 1;
            else
            {
                right = mid - 1;
            }
        }
        if(left != nums.size() and nums[left] == target){
            return left;
        }
        return -1;
    }
    int FindLastPosition(vector<int> &nums, int target){
        int left = 0, right = nums.size() - 1;
        while(left <= right){
            int mid = (left + right) / 2;
            if(nums[mid] == target)
                left = mid + 1;
            else if(nums[mid] < target)
                left = mid + 1;
            else
            {
                right = mid - 1;
            }
        }
            return right;
    }
};
// @lc code=end

